It looks like the latest TorchTitan/PyTorch broke the `aot` path (apparently because of the DispatchHandle):
```
File "/opt/pytorch/pytorch/torch/_functorch/partitioners.py", line 3649, in classify_nodes                                                     
        forward_only_graph = _extract_graph_with_inputs_outputs(                                                                                   
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                   
    File "/opt/pytorch/pytorch/torch/_functorch/partitioners.py", line 380, in _extract_graph_with_inputs_outputs                                  
        raise AssertionError(f"Node {x} was invalid, but is output")                                                                               
    AssertionError: Node tangents_2 was invalid, but is output
                                                                                                                                                   
    And from our debug instrumentation, the context around the failing node:                                                                       
                                
    DEBUG: Invalid output node: tangents_2                                                                                                         
      name=tangents_2, op=placeholder                                                                                                              
      val type=FakeScriptObject, dtype=?, shape=(2,)                                                                                               
      users: ['output']                                                                                                                            
      All 7 outputs:                                                                                                                               
        [0] add_4 op=call_function val_type=FakeTensor                                                                                             
        [1] add_37 op=call_function val_type=FakeTensor                                                                                            
        [2] add_70 op=call_function val_type=FakeTensor                                                                                            
        [3] add_103 op=call_function val_type=FakeTensor                                                                                           
        [4] add_136 op=call_function val_type=FakeTensor                                                                                           
        [5] _unsafe_view_27 op=call_function val_type=FakeTensor                                                                                   
        [6] tangents_2 op=placeholder val_type=FakeScriptObject
```
 `aot_fx_trace` still works. Let me try to add a unit test. Related: https://github.com/pytorch/torchtitan/issues/3013, https://github.com/pytorch/torchtitan/pull/3061

--- a/torchtitan/distributed/deepep/deepep.py
+++ b/torchtitan/distributed/deepep/deepep.py
@@ -241,7 +241,6 @@ torch.library.register_autograd(
 )
 
 
-@torch.compiler.disable()
 def sync_combine() -> None:
     """Synchronize the current CUDA stream with the pending combine operation.
 
@@ -251,7 +250,7 @@ def sync_combine() -> None:
     the combine to finish.
 
     torch.compile Compatibility:
-        Decorated with @torch.compiler.disable() to always run in eager mode.
+        Guarded with is_compiling() to skip in compile mode.
         This avoids issues with CUDA event operations not being traceable.
 
     Process Isolation:
@@ -277,6 +276,13 @@ def sync_combine() -> None:
     was already synced or if no combine operation is pending.
     """
     global _pending_combine_event
+    # is_compiling() is True under Dynamo (aot mode) — skip to avoid tracing
+    # through global state Dynamo can't handle. Under make_fx (aot_fx_trace),
+    # is_compiling() is False so this guard doesn't fire, but the function is
+    # still safe: _pending_combine_event is None during tracing (no real
+    # combine ran), so the body below is a no-op.
+    if torch.compiler.is_compiling():
+        return
 
     if _pending_combine_event is not None:
         _pending_combine_event.current_stream_wait()

--- a/torchtitan/distributed/deepep/hybridep.py
+++ b/torchtitan/distributed/deepep/hybridep.py
@@ -237,7 +237,7 @@ def _dispatch_fake(
     else:
         out_tokens = x.shape[0]
     hidden = x.new_empty(out_tokens, x.shape[1])
-    scores = x.new_empty(0, dtype=torch.float32)
+    scores = x.new_empty(out_tokens, dtype=torch.float32)
     tpe = x.new_empty(num_local_experts, dtype=torch.int64)
     return hidden, scores, tpe, DispatchHandle()
 
@@ -275,25 +275,34 @@ def _dispatch_backward(ctx, grad_hidden, grad_scores, grad_tpe, grad_handle):
         return None, None, None, None, None, None, None
 
     dispatch_handle = ctx.dispatch_handle
-    if dispatch_handle is None or dispatch_handle.value is None:
-        raise RuntimeError("DispatchHandle not found in dispatch backward")
-
     (topk_idx,) = ctx.saved_tensors
-    grad_x, grad_probs_dense = _buffer.combine_with_unpermute(
-        hidden=grad_hidden,
-        probs=(
-            grad_scores if grad_scores is not None and grad_scores.numel() > 0 else None
-        ),
-        handle=dispatch_handle.value,
-    )
-    grad_x = grad_x.to(ctx.input_dtype)
 
-    # grad_probs_dense is [num_tokens, num_experts]; gather back to sparse [num_tokens, top_k]
-    grad_weights = (
-        grad_probs_dense.gather(dim=1, index=topk_idx)
-        if grad_probs_dense is not None
-        else None
-    )
+    if dispatch_handle is None or dispatch_handle.value is None:
+        from torch._subclasses.fake_tensor import FakeTensor
+        from torch._subclasses.functional_tensor import FunctionalTensor
+
+        if not isinstance(grad_hidden, (FunctionalTensor, FakeTensor)):
+            raise RuntimeError("DispatchHandle not found in dispatch backward")
+        grad_x = grad_hidden.new_zeros(topk_idx.shape[0], grad_hidden.shape[-1])
+        grad_weights = grad_hidden.new_zeros(topk_idx.shape)
+    else:
+        grad_x, grad_probs_dense = _buffer.combine_with_unpermute(
+            hidden=grad_hidden,
+            probs=(
+                grad_scores
+                if grad_scores is not None and grad_scores.numel() > 0
+                else None
+            ),
+            handle=dispatch_handle.value,
+        )
+        grad_x = grad_x.to(ctx.input_dtype)
+
+        # grad_probs_dense is [num_tokens, num_experts]; gather back to sparse [num_tokens, top_k]
+        grad_weights = (
+            grad_probs_dense.gather(dim=1, index=topk_idx)
+            if grad_probs_dense is not None
+            else None
+        )
     # Gradients for: x, topk_idx, topk_weights, num_experts, non_blocking,
     #                moe_expert_capacity_factor, pad_multiple
     return grad_x, None, grad_weights, None, None, None, None
@@ -312,15 +321,22 @@ def _combine_backward(ctx, grad_combined):
     """Backward: scatter gradients via dispatch."""
     dispatch_handle = ctx.dispatch_handle
     if dispatch_handle is None or dispatch_handle.value is None:
-        raise RuntimeError("DispatchHandle not found in combine backward")
+        from torch._subclasses.fake_tensor import FakeTensor
+        from torch._subclasses.functional_tensor import FunctionalTensor
 
-    grad_x, _, _, _, _ = _buffer.dispatch_with_permute(
-        hidden=grad_combined,
-        scaling_factor=None,
-        handle=dispatch_handle.value,
-        num_permuted_tokens=ctx.num_permuted_tokens,
-        pad_multiple=ctx.pad_multiple,
-    )
+        if not isinstance(grad_combined, (FunctionalTensor, FakeTensor)):
+            raise RuntimeError("DispatchHandle not found in combine backward")
+        grad_x = grad_combined.new_zeros(
+            ctx.num_permuted_tokens, grad_combined.shape[-1]
+        )
+    else:
+        grad_x, _, _, _, _ = _buffer.dispatch_with_permute(
+            hidden=grad_combined,
+            scaling_factor=None,
+            handle=dispatch_handle.value,
+            num_permuted_tokens=ctx.num_permuted_tokens,
+            pad_multiple=ctx.pad_multiple,
+        )
     # Gradients for: x, handle, num_tokens, pad_multiple
     return grad_x, None, None, None
 
@@ -439,13 +455,14 @@ def dispatch_tokens(
     # Buffer.__init__ calls all_gather_object() which triggers aten._to_copy
     # (CUDA→CPU), a MUST_SAVE op in our SAC policy. These are infrastructure
     # ops, not model compute, and must not enter SAC's FIFO cache.
-    with _disable_current_modes():
-        get_buffer(
-            group=group,
-            hidden_dim=hidden_states.shape[1],
-            num_tokens=hidden_states.shape[0],
-            num_local_experts=num_local_experts,
-        )
+    if _buffer is None:
+        with _disable_current_modes():
+            get_buffer(
+                group=group,
+                hidden_dim=hidden_states.shape[1],
+                num_tokens=hidden_states.shape[0],
+                num_local_experts=num_local_experts,
+            )
 
     (
         hidden,

--- a/torchtitan/experiments/graph_trainer/deepseek_v3/__init__.py
+++ b/torchtitan/experiments/graph_trainer/deepseek_v3/__init__.py
@@ -20,9 +20,12 @@ def model_registry(
     flavor: str,
     attn_backend: str = "sdpa",
     moe_comm_backend: str = "standard",
+    non_blocking_capacity_factor: float | None = None,
 ) -> ModelSpec:
     base = deepseekv3_configs[flavor](
-        attn_backend=attn_backend, moe_comm_backend=moe_comm_backend
+        attn_backend=attn_backend,
+        moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     config = GraphTrainerDeepSeekV3Model.Config(
         **{f.name: getattr(base, f.name) for f in fields(base)}

--- a/torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py
+++ b/torchtitan/experiments/graph_trainer/deepseek_v3/config_registry.py
@@ -35,6 +35,18 @@ def graph_trainer_deepseek_v3_debugmodel_ep() -> GraphTrainer.Config:
     return config
 
 
+def graph_trainer_deepseek_v3_debugmodel_hybridep() -> GraphTrainer.Config:
+    config = to_graph_trainer_config(deepseek_v3_debugmodel_ep(), model_registry)
+    config.compile = GraphTrainerCompileConfig(enable=True)
+    config.model_spec = model_registry(
+        "debugmodel",
+        moe_comm_backend="hybridep",
+        non_blocking_capacity_factor=1.0,
+    )
+    config.parallelism.expert_parallel_comm_backend = "hybridep"
+    return config
+
+
 def graph_trainer_deepseek_v3_debugmodel_flex_attn() -> (GraphTrainer.Config):
     config = to_graph_trainer_config(deepseek_v3_debugmodel_flex_attn(), model_registry)
     config.compile = GraphTrainerCompileConfig(enable=True)

--- a/torchtitan/experiments/graph_trainer/deepseek_v3/parallelize.py
+++ b/torchtitan/experiments/graph_trainer/deepseek_v3/parallelize.py
@@ -84,6 +84,10 @@ def parallelize_deepseekv3(
         # config in Trainer.Config.__post_init__; Module.parallelize applies it.
         model.parallelize(tp_mesh)
 
+    comm_backend = parallelism.expert_parallel_comm_backend
+    if comm_backend == "hybridep":
+        from torchtitan.distributed.deepep import hybridep  # noqa: F401
+
     if parallel_dims.tp_enabled or parallel_dims.ep_enabled:
         apply_moe_ep_tp(
             model,
@@ -92,12 +96,32 @@ def parallelize_deepseekv3(
             etp_mesh=parallel_dims.get_optional_mesh("etp"),
             ep_etp_mesh=parallel_dims.get_optional_mesh(["ep", "etp"]),
             enable_sp=parallelism.enable_sequence_parallel,
+            comm_backend=comm_backend,
         )
 
     # Apply simple_fsdp unconditionally. The `fsdp` mesh always exists with a
     # real backend (see ParallelDims._mesh_exist), even at degree 1, so that
     # MixedPrecisionPolicy's param_dtype cast still applies in single-GPU runs.
     model = apply_simple_fsdp(model, parallel_dims=parallel_dims, training=training)
+
+    if comm_backend == "hybridep" and parallel_dims.ep_enabled:
+        from torchtitan.distributed.deepep.hybridep import get_buffer
+
+        ep_group = parallel_dims.get_mesh("ep").get_group()
+        moe_config = next(l.moe for l in model.config.layers if l.moe is not None)
+        num_local_experts = moe_config.num_experts // parallel_dims.ep
+        hidden_dim = model.config.dim
+        num_tokens = training.local_batch_size * training.seq_len
+        get_buffer(
+            group=ep_group,
+            hidden_dim=hidden_dim,
+            num_tokens=num_tokens,
+            num_local_experts=num_local_experts,
+        )
+        logger.info(
+            f"Pre-initialized HybridEP buffer (hidden_dim={hidden_dim}, "
+            f"num_tokens={num_tokens}, num_local_experts={num_local_experts})"
+        )
 
     # Apply compilation based on mode
     model = apply_compile(

--- a/torchtitan/models/deepseek_v3/__init__.py
+++ b/torchtitan/models/deepseek_v3/__init__.py
@@ -172,7 +172,7 @@ def _build_dsv3_layers(
     score_before_experts: bool = False,
     attn_backend: str,
     moe_comm_backend: str,
-    non_blocking_capacity_factor: float | None = None,
+    non_blocking_capacity_factor: float | None,
 ) -> list[TransformerBlock.Config]:
     """Build the list of per-layer TransformerBlock configs.
 
@@ -255,6 +255,7 @@ def _build_dsv3_layers(
 def _debugmodel(
     attn_backend: str,
     moe_comm_backend: str,
+    non_blocking_capacity_factor: float | None,
 ) -> DeepSeekV3Model.Config:
     dim = 256
     n_layers = 6
@@ -287,6 +288,7 @@ def _debugmodel(
         score_before_experts=False,
         attn_backend=attn_backend,
         moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     return DeepSeekV3Model.Config(
         vocab_size=vocab_size,
@@ -318,6 +320,7 @@ def _debugmodel(
 def _16b(
     attn_backend: str,
     moe_comm_backend: str,
+    non_blocking_capacity_factor: float | None,
 ) -> DeepSeekV3Model.Config:
     dim = 2048
     n_layers = 27
@@ -350,6 +353,7 @@ def _16b(
         score_before_experts=False,
         attn_backend=attn_backend,
         moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     return DeepSeekV3Model.Config(
         vocab_size=vocab_size,
@@ -381,6 +385,7 @@ def _16b(
 def _236b(
     attn_backend: str,
     moe_comm_backend: str,
+    non_blocking_capacity_factor: float | None,
 ) -> DeepSeekV3Model.Config:
     dim = 5120
     n_layers = 60
@@ -417,6 +422,7 @@ def _236b(
         score_before_experts=False,
         attn_backend=attn_backend,
         moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     return DeepSeekV3Model.Config(
         vocab_size=vocab_size,
@@ -448,6 +454,7 @@ def _236b(
 def _671b(
     attn_backend: str,
     moe_comm_backend: str,
+    non_blocking_capacity_factor: float | None,
 ) -> DeepSeekV3Model.Config:
     dim = 7168
     n_layers = 61
@@ -485,6 +492,7 @@ def _671b(
         score_before_experts=False,
         attn_backend=attn_backend,
         moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     return DeepSeekV3Model.Config(
         vocab_size=vocab_size,
@@ -525,11 +533,13 @@ def model_registry(
     flavor: str,
     attn_backend: str = "sdpa",
     moe_comm_backend: str = "standard",
+    non_blocking_capacity_factor: float | None = None,
     quantization: list[QuantizationConverter.Config] | None = None,
 ) -> ModelSpec:
     config = deepseekv3_configs[flavor](
         attn_backend=attn_backend,
         moe_comm_backend=moe_comm_backend,
+        non_blocking_capacity_factor=non_blocking_capacity_factor,
     )
     if quantization is not None:
         for q in quantization:

